PR 1

VladX opened the pull request.

VladX committed:
Added naive blockmatching algorithm (just to get acquainted with the code base, this algo shouldn't be considered seriously).
Also, added boilerplate for pcaflow.

VladX committed:
Added style rules for clang-format

VladX committed:
Added flow estimation using DCT basis

VladX committed:
Some improvements

VladX committed:
Boosted accuracy

VladX committed:
Boosted accuracy a little bit more

VladX committed:
Preserve edges

VladX committed:
Added support for prior (covariance matrix)

VladX committed:
Added OpenCL support

ethanrublee commented on modules/optflow/include/opencv2/optflow/pcaflow.hpp:
```diff
@@ -0,0 +1,109 @@
+               For Open Source Computer Vision Library
+                       (3-clause BSD License)
+
+Copyright (C) 2013, OpenCV Foundation, all rights reserved.
```
Consider updating the date here.

ethanrublee commented on modules/optflow/include/opencv2/optflow/pcaflow.hpp:
```diff
@@ -0,0 +1,109 @@
+{
+namespace optflow
+{
+namespace pcaflow
```
I wonder if you need the additional namespace.  cv::optflow might be sufficient for all the algorithms.

VladX replied:
This namespace contains only "Prior" class. Without this namespace I experienced compilation errors related to Python module... I didn't know how to fix, so I just framed the class with the namespace.

ethanrublee replied:
You can mangle the Prior class to PCAPrior, or similar.

ethanrublee commented on modules/optflow/include/opencv2/optflow/pcaflow.hpp:
```diff
@@ -0,0 +1,109 @@
+namespace pcaflow
+{
+
+class Prior
```
A comment about the purpose of this type would help the reader.

ethanrublee commented on modules/optflow/src/block_matching.cpp:
```diff
@@ -0,0 +1,143 @@
+  void collectGarbage();
+};
+
+inline float OpticalFlowBlockMatching::submatrixAbsDiff( int x0, int y0, const Mat &I0, int x1, int y1,
```
It seems this function can be written in terms of ROIs, might be less code to maintain, and can take advantage of any already optimized/tested code.

cv::Rect valid_roi0, valid_roi1; // compute these based on x0,y0,x1,y1, and the image dimensions.
cv::absDiff(I0(valid_roi0), I1(valid_roi1), diff);
float cost = cv::sum(diff) + area_of_invalid_region;

ethanrublee commented on modules/optflow/src/block_matching.cpp:
```diff
@@ -0,0 +1,143 @@
+
+void OpticalFlowBlockMatching::calc( InputArray I0, InputArray I1, InputOutputArray flow_out )
+{
+  CV_Assert( I0.channels() == 3 );
```
Should this work for grayscale as well?

VladX replied:
I think this "block matching" shouldn't be here at all :) I wrote it for a testing purpose to get acquainted with the existing optflow codebase. This is not supposed to be used by anyone else.

ethanrublee replied:
Ok, yeah, makes sense.  Suggest removal :)

ethanrublee commented on modules/optflow/src/learn_prior.py:
```diff
@@ -0,0 +1,123 @@
+import os, sys
```
Consider adding file comment, to give some idea on the purpose of this script.

Also suggest putting this through a PEP8 autoformatter, like yapf.

ethanrublee replied:
An example of how to use this script, so that anyone might be able to reproduce the results would be great.  Include where/how to get the data, and sequence of steps necessary to construct the prior file.

ethanrublee commented on modules/optflow/src/pcaflow.cpp:
```diff
@@ -0,0 +1,499 @@
+  CV_Assert( occlusionsThreshold > 0 );
+}
+
+inline float eDistSq( const Point2f &p1, const Point2f &p2 )
```
These functions (eDistSq, eNormSq) must be implemented somewhere else in opencv.  Consider not rolling your own.

ethanrublee commented on modules/optflow/src/pcaflow.cpp:
```diff
@@ -0,0 +1,499 @@
+
+inline float eNormSq( const Point2f &v ) { return v.x * v.x + v.y * v.y; }
+
+template <typename T> static inline int mathSign( T val ) { return ( T( 0 ) < val ) - ( val < T( 0 ) ); }
```
It seems like this "sgn" or "signum" function should be somewhere in opencv's core.  I wonder if there are any existing implementations in the library. I can't find one after a quick search; but didn't look to hard.

VladX committed:
Post-review fixes

VladX committed:
Merge recent changes

ethanrublee commented:
Looks like the diff is no longer clean.  Suggest updating the master branch on this repo to reflect upstream.

ethanrublee commented on modules/tracking/src/onlineMIL.cpp:
```diff
@@ -42,6 +42,8 @@
 #include "precomp.hpp"
 #include "opencv2/tracking/onlineMIL.hpp"
 
+#define  sign(s)  ((s > 0 ) ? 1 : ((s<0) ? -1 : 0))
```
@VladX Ha, here is another signum function :)

ethanrublee commented on modules/optflow/src/pcaflow.cpp:
```diff
@@ -0,0 +1,486 @@
+
+template <typename T> static inline int mathSign( T val ) { return ( T( 0 ) < val ) - ( val < T( 0 ) ); }
+
+static inline void symOrtho( double a, double b, double &c, double &s, double &r )
```
Some comments for symOrtho and solveLSQR would be helpful.  What is the algorithm, what's the reference used, what are the arguments? I realize these are internal functions, but still easier for maintenance to jot this stuff down.

ethanrublee commented on modules/optflow/src/pcaflow.cpp:
```diff
@@ -0,0 +1,486 @@
+  {
+    const unsigned missingPoints = maxFeatures - features.size();
+    const unsigned blockSize = sqrt( (float)size.area() / missingPoints );
+    for ( int x = blockSize / 2; x < size.width; x += blockSize )
```
Should you add these features anyway?  It seems like it might be easy to have frames where there are a high concentration of features in one area of the image ( found through goodFeaturesToTrack) which satisfies your maxFeatures threshold.  Is the goal to have a minimum density of features?

ethanrublee commented:
Ok, was able to run ./example_optflow_optical_flow_evaluation on some data.

I haven't had a chance to run the benchmarking script, which produces charts. I'd love to see a readme that explains how to get data and what the expected results are for pcaflow, and the other algorithms.

Still would like to do review another pass over the core of the algorithm, probably tomorrow.

VladX committed:
Added some explanatory comments

VladX commented:
`python opencv_contrib/modules/optflow/samples/optical_flow_benchmark.py ./bin/example_optflow_optical_flow_evaluation -f ~/flow_datasets/ -a farneback,DISflow,pcaflow,tvl1,deepflow -d middlebury`

I ran the new benchmarking script and this is what I got:
![evaluation_results_middlebury](https://cloud.githubusercontent.com/assets/74066/16011184/04d7f1cc-318d-11e6-83f7-55efd85db779.png)

Unfortunately, the tool seems to fail if I try to evaluate more than 5 algorithms.

You need to unpack an archive with the middlebury dataset and ground truth flows to the ~/flow_datasets/

VladX committed:
Passing of the prior file to the evaluation tool

VladX committed:
Minor fixes

ethanrublee commented on modules/optflow/src/.clang-format:
```diff
@@ -0,0 +1,65 @@
+---
+Language:        Cpp
```
Suggest removing this file.  It seems like cruft, I don't see a large precedence for clang-format files in opencv.

ethanrublee commented on modules/optflow/src/pcaflow.cpp:
```diff
@@ -0,0 +1,508 @@
+    solveLSQR( A, b1, w1, dampingFactor * size.area() );
+    solveLSQR( A, b2, w2, dampingFactor * size.area() );
+  }
+  Mat flowSmall( ( size / 8 ) * 2, CV_32FC2 );
```
What's the rationale behind this small size?  perhaps a comment, and suggest (size/4)

ethanrublee commented on modules/optflow/src/pcaflow.cpp:
```diff
@@ -0,0 +1,508 @@
+ *   s -- sine(theta)
+ *   r -- two-norm of [a; b]
+ */
+static inline void symOrtho( double a, double b, double &c, double &s, double &r )
```
Suggest adding unit test for symOrtho.

ethanrublee commented on modules/optflow/src/pcaflow.cpp:
```diff
@@ -0,0 +1,508 @@
+ * Output:
+ *   x -- approximate solution
+ */
+static void solveLSQR( const Mat &A, const Mat &b, OutputArray xOut, const double damp = 0.0,
```
Add unit test for solveLSQR.

VladX replied:
What is the best way to add a unit test for internal function? Can't find any example in the existing opencv code.

ethanrublee commented on modules/optflow/src/pcaflow.cpp:
```diff
@@ -0,0 +1,508 @@
+{
+  int m = A.size().height;
+  int n = A.size().width;
+  CV_Assert( m == b.size().height );
```
Are you checking all of the preconditions here?

ethanrublee commented on modules/optflow/src/pcaflow.cpp:
```diff
@@ -0,0 +1,508 @@
+  predictedFeatures.resize( j );
+}
+
+static inline void _cpu_fillDCTSampledPoints( float *row, const Point2f &p, const Size &basisSize, const Size &size )
```
Advise using anonymous namespace for these functions to hide linkage, rather than the static keyword.  More idiomatic c++.

ethanrublee commented on modules/optflow/src/pcaflow.cpp:
```diff
@@ -0,0 +1,508 @@
+static void applyCLAHE( UMat &img )
+{
+  Ptr<CLAHE> clahe = createCLAHE();
+  clahe->setClipLimit( 14 );
```
where did this magic number come from?

ethanrublee commented:
So I think this looks pretty good for a first pass.  Suggest adding unit tests, and submitting pull request to the main repository.

I have not trained the prior on my box.

I did notice that it seemed to not work as well for higher resolution input, like 1200x900.

VladX committed:
PR fixes

VladX commented:
> I did notice that it seemed to not work as well for higher resolution input, like 1200x900.

That's because the size of DCT basis is fixed and independent from the actual image resolution.
During a calculation of prior the size of DCT basis need to be known in advance. Thus it must be fixed. Unpleasant but necessary limitation of the PCAFlow algorithm.

VladX committed:
Fixed build on VS

VladX committed:
Fixing some compiler warnings

VladX committed:
Fixed binary file reading on Windows, math constants and OpenCL kernel

VladX committed:
Merge upstream

VladX committed:
Added a note about key differences

VladX committed:
Added training part of the Global Patch Collider

VladX committed:
Added new method for training forest

VladX committed:
Merge https://github.com/Itseez/opencv_contrib into optflow

VladX committed:
Fixing GCC warning

VladX committed:
Fixing VS warning

VladX committed:
Fixing signed vs unsigned warning

VladX committed:
Minor fixes

VladX committed:
Start adding correspondence finding

VladX committed:
Fixed OpenCL check behavior

VladX committed:
Added finding of correspondences using GPC

VladX committed:
New evaluation tool for GPC

VladX committed:
Changed default parameters

VladX committed:
Display ground truth in the evaluation tool

VladX committed:
Added training tool for MPI Sintel dataset

VladX committed:
Added the training tool for Middlebury dataset

VladX committed:
Added some OpenCL optimization

VladX committed:
Added explanatory notes

VladX committed:
Minor improvements: time measurements + little ocl optimization

VladX committed:
Merge branch 'gpc' into optflow

VladX committed:
Added demos

VladX committed:
Fixed warnings

VladX committed:
Make parameter struct assignable

VladX committed:
Fix warning

VladX committed:
Proper command line argument usage

VladX committed:
Prettified training tool, added parameters

VladX committed:
Fixed VS warning

VladX committed:
Fixed VS warning

VladX committed:
Using of compressed forest.yml.gz files by default to save space

VladX committed:
Added OpenCL flag to the evaluation tool

VladX committed:
Updated documentation

VladX committed:
Major speed and memory improvements:
1) Added new (optional) type of patch descriptors which are much faster. Retraining with option --descriptor-type=1 is required.
2) Got rid of hash table for descriptors, less memory usage.

VladX committed:
Fixed various floating point errors related to precision.
SIMD for dot product, forest traversing is a little bit faster now.

VladX committed:
Tolerant floating point comparison

VladX committed:
Triplets

VladX committed:
Added comment

VladX committed:
Choosing negative sample among nearest neighbors

VladX committed:
Fix warning

VladX committed:
Usage of parallel_for_() in critical places. Performance improvments.

VladX committed:
Simulated annealing heuristic

VladX committed:
Moved OpenCL kernel to separate file

VladX committed:
Moved implementation to source file

VladX committed:
Added basic accuracy tests for GPC and PCAFlow

VladX committed:
Fixing warnings

VladX committed:
Test accuracy constraints were too strict

VladX committed:
Test accuracy constraints were too strict

VladX committed:
Make tests more lightweight
